documentation/OpenMP-4.5-grammar.txt

```diff
@@ -1,114 +1,396 @@
+
+2.16 declare-reduction -> DECLARE REDUCTION (reduction-identifier : type-list : combiner)
 
 # Clauses
-2.5 proc_bind -> PROC_BIND ( MASTER | CLOSE | SPREAD )
-2.5 num_threads -> NUM_THREADS(scalar-int-expr)
-2.7.1 schedule -> SCHEDULE ([sched-modifier] [, sched-modifier]:]
-                  kind[, chunk_size])
+2.5 proc-bind-clause -> PROC_BIND (MASTER | CLOSE | SPREAD)
+
+2.5 num-threads-clause -> NUM_THREADS (scalar-int-expr)
+
+2.7.1 schedule-clause -> SCHEDULE ([sched-modifier] [, sched-modifier]:]
+                                   kind[, chunk_size])
+
 2.7.1 kind -> STATIC | DYNAMIC | GUIDED | AUTO | RUNTIME
+
 2.7.1 sched-modifier -> MONOTONIC | NONMONOTONIC | SIMD
+
 2.7.1 chunk_size -> scalar-int-expr
-2.7.1 collapse -> COLLAPSE(scalar-constant)
-2.7.1 ordered -> ORDERED [ (scalar-constant) ]
-2.7.1 nowait -> NOWAIT
-2.8.1 aligned -> ALIGNED ( variable-name-list [ : scalar-constant])
-2.8.1 safelen -> SAFELEN ( scalar-constant )
-2.8.1 simdlen -> SIMDLEN ( scalar-contant )
-2.8.2 uniform -> UNIFORM (dummy-arg-name-list)
-2.8.2 inbranch -> INBRANCH
-2.8.2 notinbranch -> NOTINBRANCH
-2.13.9 depend -> DEPEND (((IN | OUT | INOUT) : variable-name-list) |
-                          SOURCE |
-                          SINK : vec)
-         vec -> iterator [ +/- scalar-int-expr:scalar-int-expr],..., iterator[...]
-2.9.2 num_tasks -> NUM_TASKS (scalar-int-expr)
-2.9.2 grainsize -> GRAINSIZE (scalar-int-expr)
-2.9.2 nogroup -> NOGROUP
-2.9.2 untied -> UNTIED
-2.9.2 priority -> PRIORITY (scalar-int-expr)
-2.9.2 mergeable -> MERGEABLE
-2.9.2 final -> FINAL (scalar-int-expr)
-2.10.1 use_device_ptr -> USE_DEVICE_PTR (variable-name-list)
-2.10.1 device -> DEVICE (scalar-integer-expr)
-2.10.5 to -> TO (variable-name-list)
-2.10.5 from -> FROM (variable-name-list)
-2.10.6 link -> LINK (variable-name-list)
-2.10.7 num_teams -> NUM_TEAMS (scalar-integer-expr)
-2.10.7 thread_limit -> THREAD_LIMIT (scalar-integer-expr)
-2.10.8 dist_schedule -> DIST_SCHEDULE (STATIC [ , chunk_size])
-2.12 if -> IF ([ directive-name-modifier ':' ] scalar-logical-expr)
-2.15.3.1 default -> DEFAULT (PRIVATE | FIRSTPRIVATE | SHARED | NONE)
-2.15.3.2 shared -> SHARED (variable-name-list)
-2.15.3.3 private -> PRIVATE (variable-name-list)
-2.15.3.4 firstprivate -> FIRSTPRIVATE (variable-name-list)
-2.15.3.5 lastprivate -> LASTPRIVATE (variable-name-list)
-2.15.3.6 reduction -> REDUCTION(reduction-identifier: variable-name-list)
-2.15.3.7 linear -> LINEAR(linear-list [:linear-step])
-           linear-list -> list | modifier(list)
-           modifier -> REF | VAL | UVAL
-2.15.4.1 copyin -> COPYIN (variable-name-list)
-2.15.4.2 copyprivate -> COPYPRIVATE (variable-name-list)
-2.15.5.1 map -> MAP ([[ALWAYS[,]] (TO | FROM | TOFROM | ALLOC | RELEASE |
+
+2.7.1 collapse-clause -> COLLAPSE (scalar-constant)
+
+2.7.1 ordered-clause -> ORDERED [(scalar-constant)]
+
+2.7.1 nowait-clause -> NOWAIT
+
+2.8.1 aligned-clause -> ALIGNED (variable-name-list[ : scalar-constant])
```

Comment:
The specification states that the "alignment" should be a constant positive integer expression.  Is "scalar-constant" the correct non-terminal to specify this?

Reply:
Yes, we don't have `ScalarPositiveIntConstantExpr` defined in the parser. The check can be deferred to Semantics.

Reply:
Fortran doesn't have a constraint for positive values in its grammar.  It does have constraints for scalar, constant, and integer, so you can use those at least, and then check for a positive value in semantics.